xernel/kernel/src/arch/amd64/interrupts/mod.rs

```diff
@@ -0,0 +1,199 @@
+pub mod dpc;
+pub mod dpc_queue;
+pub mod idt;
+pub mod ipl;
+
+use crate::arch::amd64::apic::APIC;
+use crate::arch::amd64::{ports::outb, read_cr2};
+use crate::sched::context::TrapFrame;
+use core::arch::asm;
+use core::sync::atomic::{compiler_fence, Ordering};
+use idt::{IRQHandler, IDT_ENTRIES};
+use ipl::IPL;
+
+use self::ipl::{get_spl, raise_spl, set_ipl};
+
+use super::apic::apic_spurious_interrupt;
+use libxernel::sync::{Spinlock, SpinlockIRQ};
+
+static INTERRUPT_HANDLERS: SpinlockIRQ<[IRQHandler; IDT_ENTRIES]> = SpinlockIRQ::new([IRQHandler::None; IDT_ENTRIES]);
+
+pub fn init() {
+    idt::init();
+
+    let mut handlers = INTERRUPT_HANDLERS.lock();
+
+    handlers[0xD] = IRQHandler::Handler(general_fault_handler);
+    handlers[0xE] = IRQHandler::Handler(page_fault_handler);
+    handlers[0x8] = IRQHandler::Handler(double_fault_handler);
+    handlers[0xF0] = IRQHandler::Handler(apic_spurious_interrupt);
+}
+
+#[no_mangle]
+extern "sysv64" fn generic_interrupt_handler(isr: usize, ctx: *mut TrapFrame) {
+    let mut ipl = IPL::from(isr >> 4);
+
+    if (ipl as u8) < (get_spl() as u8) {
+        panic!("IPL not less or equal");
+    }
+
+    ipl = raise_spl(ipl);
+
+    let handlers = INTERRUPT_HANDLERS.lock();
+
+    let ctx = unsafe { &mut *ctx };
+
+    enable();
+
+    match &handlers[isr] {
+        IRQHandler::Handler(handler) => {
+            let handler = *handler;
+            handlers.unlock();
+            handler(ctx);
+        }
+
+        IRQHandler::None => panic!("unhandled interrupt {}", isr),
+    }
+
+    if isr > 32 {
+        APIC.eoi();
+    }
+
+    disable();
+
+    set_ipl(ipl);
+}
+
+#[inline]
+pub fn enable() {
+    compiler_fence(Ordering::Release);
+    unsafe {
+        asm!("sti", options(nomem, nostack));
+    }
+}
+
+#[inline]
+pub fn disable() {
+    compiler_fence(Ordering::Acquire);
+    unsafe {
+        asm!("cli", options(nomem, nostack));
+    }
+}
+
+// FIXME: Find solution for multi-core usage
+pub fn allocate_vector(ipl: IPL) -> Option<u8> {
+    static FREE_VECTORS_FOR_IPL: Spinlock<[u8; 16]> = Spinlock::new([
+        0x0 << 4,
+        0x1 << 4,
+        0x2 << 4,
+        0x3 << 4,
+        0x4 << 4,
+        0x5 << 4,
+        0x6 << 4,
+        0x7 << 4,
+        0x8 << 4,
+        0x9 << 4,
+        0xA << 4,
+        0xB << 4,
+        0xC << 4,
+        0xD << 4,
+        0xE << 4,
+        0xF << 4,
+    ]);
+
+    if (ipl as u8) > 15 {
+        return None;
+    }
+
+    let base_vector = (ipl as u8) << 4;
+
+    let mut free_vectors = FREE_VECTORS_FOR_IPL.lock();
+
+    let next_free_vector = free_vectors[ipl as usize];
+
+    if next_free_vector > base_vector + 15 {
+        return None;
+    }
+
+    free_vectors[ipl as usize] += 1;
+
+    Some(next_free_vector)
+}
```

Comment:
This function does not check whether the assigned vector was already set in the ``init`` function

Reply:
Currently not used. Will be rewritten when actually used